src/control/probe.rs

```diff
@@ -0,0 +1,469 @@
+///
+/// A triangular probe has us jump by 1 more group every time. So first we
+/// jump by 1 group (meaning we just continue our linear scan), then 2 groups
+/// (skipping over 1 group), then 3 groups (skipping over 2 groups), and so on.
+///
+/// Proof that the probe will visit every group in the table:
+/// <https://fgiesen.wordpress.com/2015/02/22/triangular-numbers-mod-2n/>
+#[derive(Clone)]
+pub(crate) struct Probe {
+    /// Current position in sequence.
+    pos: usize,
+
+    /// Offset from current position to next group.
+    stride: usize,
+
+    /// Pointer to control data.
+    ctrl: NonNull<Tag>,
+
+    /// Number of buckets, minus one.
+    bucket_mask: usize,
+
+    /// Tag containing the hash of the item.
+    hash_tag: Tag,
+
+    /// Initial hash, for debugging.
+    #[cfg(debug_assertions)]
+    hash: u64,
+
+    /// Current iteration, for debugging.
+    #[cfg(debug_assertions)]
+    iter: usize,
+}
+impl Probe {
+    /// Creates a new probe for a table for a given hash.
+    ///
+    /// # Safety
+    ///
+    /// The given bucket mask must be equal to the power-of-two number of buckets minus one,
+    /// and the given control pointer must be a valid hash table.
+    pub(crate) unsafe fn new(ctrl: NonNull<Tag>, bucket_mask: usize, hash: u64) -> Probe {
+        debug_assert!((bucket_mask + 1).is_power_of_two());
+
+        Probe {
+            // On 32-bit platforms we simply ignore the higher hash bits.
+            //
+            // We need to apply the bucket mask now in the rare case where it
+            // would fall greater than `usize::MAX - Group::WIDTH`, since it
+            // could overflow.
+            pos: (hash as usize) & bucket_mask,
+            hash_tag: Tag::full(hash),
+            stride: 0,
+            ctrl,
+            bucket_mask,
+            #[cfg(debug_assertions)]
+            hash,
+            #[cfg(debug_assertions)]
+            iter: 0,
+        }
+    }
+
+    /// Gets the inner slice of groups.
+    fn control_slice(&self) -> &[Tag] {
+        // SAFETY: We always allocate this many control tags.
+        unsafe { slice::from_raw_parts(self.ctrl.as_ptr(), self.bucket_mask + 1 + Group::WIDTH) }
+    }
+
+    /// Verifies that the control slice is valid.
+    fn assert_at_least_one_empty(&self) {
+        let control = self.control_slice();
+        debug_assert!(
+            control.iter().any(|tag| tag.is_empty()),
+            "control slice was completely full: {:x?}",
+            control
+        );
+    }
+
+    /// Finds an empty slot to insert something.
+    ///
+    /// Although not a safety violation, this method *will* loop infinitely if
+    /// the table does not contain an empty slot. You have been warned.
+    pub(crate) fn empty(mut self) -> usize {
+        self.assert_at_least_one_empty();
+        loop {
+            // SAFETY: We always return an item from the iterator.
+            let group_probe = unsafe { self.next().unwrap_unchecked() };
```

Comment:
Rather than using `unwrap_unchecked` everywhere, it would be better to not use the `Iterator` trait here and just have an inherent `next` method.